cpg_workflows/stages/realign_genotype_with_dragen.py

```diff
@@ -0,0 +1,329 @@
+
+        prepare_ica_job: PythonJob = get_batch().new_python_job(
+            name='PrepareIcaForDragenAnalysis',
+            attributes=(self.get_job_attrs() or {}) | {'tool': 'ICA'},
+        )
+        prepare_ica_job.image(image=image_path('cpg_workflows'))
+
+        outputs = self.expected_outputs(sequencing_group=sequencing_group)
+        output_fids = prepare_ica_job.call(
+            prepare_ica_for_analysis.run,
+            ica_analysis_output_folder=config_retrieve(['ica', 'data_prep', 'output_folder']),
+            api_root=ICA_REST_ENDPOINT,
+            sg_name=sequencing_group.name,
+            bucket_name=bucket_name,
+        ).as_json()
+
+        get_batch().write_output(
+            output_fids,
+            str(outputs),
+        )
+        return self.make_outputs(
+            target=sequencing_group,
+            data=outputs,
+            jobs=prepare_ica_job,
+        )
+
+
+@stage
+class UploadDataToIca(SequencingGroupStage):
+    def expected_outputs(self, sequencing_group: SequencingGroup) -> cpg_utils.Path:
+        sg_bucket: cpg_utils.Path = sequencing_group.dataset.prefix()
+        return sg_bucket / GCP_FOLDER_FOR_ICA_PREP / f'{sequencing_group.name}_fids.json'
+
+    def queue_jobs(self, sequencing_group: SequencingGroup, inputs: StageInput) -> StageOutput | None:
+        upload_job = get_batch().new_bash_job(
+            name='UploadDataToIca',
+            attributes=(self.get_job_attrs() or {}) | {'tool': 'ICA'},
+        )
+        upload_folder = config_retrieve(['ica', 'data_prep', 'upload_folder'])
+        bucket: str = get_path_components_from_gcp_path(str(sequencing_group.cram))['bucket']
+
+        upload_job.image(image=image_path('ica'))
+        upload_job.storage(calculate_needed_storage(cram=str(sequencing_group.cram)))
+        output = self.expected_outputs(sequencing_group=sequencing_group)
+        authenticate_cloud_credentials_in_job(upload_job)
+
+        # Check if the CRAM already exists in ICA before uploading. If it exists, just return the ID for the CRAM and CRAI
+        upload_job.command(
+            f"""
+            {ICA_CLI_SETUP}
+            cram_status=$(icav2 projectdata list --parent-folder /{bucket}/{upload_folder}/{sequencing_group.name}/ --data-type FILE --file-name {sequencing_group.name}.cram --match-mode EXACT -o json | jq -r '.items[].id.details.status')
+            if [[ $cram_status != "AVAILABLE" ]]
+            then
+                mkdir {sequencing_group.name}
+                gcloud storage cp {sequencing_group.cram} .
+                gcloud storage cp {sequencing_group.cram}.crai .
+                icav2 projectdata upload {sequencing_group.name}.cram /{bucket}/{upload_folder}/{sequencing_group.name}/
+                icav2 projectdata upload {sequencing_group.name}.cram.crai /{bucket}/{upload_folder}/{sequencing_group.name}/
+            fi
```

Comment:
```suggestion
        cram_group = get_batch().read_input_group(
            cram=str(sequencing_group.cram.path), 
            crai=str(sequencing_group.cram.index_path),
        )
        
        # Check if the CRAM already exists in ICA before uploading. If it exists, just return the ID for the CRAM and CRAI
        upload_job.command(
            f"""
            {ICA_CLI_SETUP}
            cram_status=$(icav2 projectdata list --parent-folder /{bucket}/{upload_folder}/{sequencing_group.name}/ --data-type FILE --file-name {sequencing_group.name}.cram --match-mode EXACT -o json | jq -r '.items[].id.details.status')
            if [[ $cram_status != "AVAILABLE" ]]
            then
                icav2 projectdata upload {cram_group.cram} /{bucket}/{upload_folder}/{sequencing_group.name}/
                icav2 projectdata upload {cram_group.crai} /{bucket}/{upload_folder}/{sequencing_group.name}/
            fi
```

just a suggestion to delegate localisation to hail batch, which would make it easier to drop the gcloud cli in this job image later

Reply:
This diff is a bit messy given that it includes the unchanged lines in between. It's the read_input_group to create the local files from cram/crai, and dropping the two gcloud cp lines

Reply:
So I know that it doesn't take a huge amount of time to load data from GCP to the VM, but the current solution only does that if the CRAM is not already present in ICA, instead of always. Maybe I'm solving for a problem that won't really come up, but 🤷